File: sections/72_master_flow_convex_mixed/problems/b_bounded_exact_shipment/solution.py
import collections
import sys


class MinCostFlow:
    def __init__(self, n):
        self.graph = [[] for _ in range(n)]

    def add_edge(self, u, v, capacity, cost):
        self.graph[u].append([v, len(self.graph[v]), capacity, cost])
        self.graph[v].append([u, len(self.graph[u]) - 1, 0, -cost])

    def flow(self, source, sink, required):
        sent = total_cost = 0
        n = len(self.graph)
        while sent < required:
            distance = [None] * n
            parent = [None] * n
            queued = [False] * n
            queue = collections.deque([source])
            distance[source] = 0
            queued[source] = True
            while queue:
                u = queue.popleft()
                queued[u] = False
                for index, edge in enumerate(self.graph[u]):
                    v, _, capacity, cost = edge
                    if not capacity:
                        continue
                    candidate = distance[u] + cost
                    if distance[v] is not None and candidate >= distance[v]:
                        continue
                    distance[v] = candidate
                    parent[v] = (u, index)
                    if not queued[v]:
                        queued[v] = True
                        queue.append(v)
            if distance[sink] is None:
                break
            add = required - sent
            vertex = sink
            while vertex != source:
                u, index = parent[vertex]
                add = min(add, self.graph[u][index][2])
                vertex = u
            vertex = sink
            while vertex != source:
                u, index = parent[vertex]
                edge = self.graph[u][index]
                total_cost += add * edge[3]
                edge[2] -= add
                self.graph[vertex][edge[1]][2] += add
                vertex = u
            sent += add
        return sent, total_cost


def main():
    data = list(map(int, sys.stdin.buffer.read().split()))
    n, m, exact = data[:3]
    source, sink = data[3] - 1, data[4] - 1
    at = 5
    network = MinCostFlow(n + 2)
    balance = [0] * n
    mandatory_cost = 0

    def add_bounded(u, v, low, high, cost):
        nonlocal mandatory_cost
        balance[u] -= low
        balance[v] += low
        mandatory_cost += low * cost
        network.add_edge(u, v, high - low, cost)

    for _ in range(m):
        u, v, low, high, cost = data[at:at + 5]
        at += 5
        add_bounded(u - 1, v - 1, low, high, cost)
    add_bounded(sink, source, exact, exact, 0)

    super_source, super_sink = n, n + 1
    required = 0
    for vertex, value in enumerate(balance):
        if value > 0:
            network.add_edge(super_source, vertex, value, 0)
            required += value
        elif value < 0:
            network.add_edge(vertex, super_sink, -value, 0)

    sent, extra_cost = network.flow(super_source, super_sink, required)
    if sent != required:
        print("IMPOSSIBLE")
    else:
        print(mandatory_cost + extra_cost)


if __name__ == '__main__':
    main()

File: sections/72_master_flow_convex_mixed/problems/b_bounded_exact_shipment/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
const ll INF = (1LL << 62);

struct MinCostFlow {
    struct Edge { int to, rev, cap; ll cost; };
    vector<vector<Edge>> graph;

    explicit MinCostFlow(int n) : graph(n) {}

    void add_edge(int u, int v, int cap, ll cost) {
        int a = graph[u].size();
        int b = graph[v].size();
        graph[u].push_back({v, b, cap, cost});
        graph[v].push_back({u, a, 0, -cost});
    }

    pair<int, ll> flow(int source, int sink, int required) {
        int sent = 0;
        ll cost = 0;
        int n = graph.size();
        vector<ll> distance(n);
        vector<int> parent_v(n), parent_e(n);

        while (sent < required) {
            fill(distance.begin(), distance.end(), INF);
            vector<char> queued(n, false);
            queue<int> queue;
            distance[source] = 0;
            queue.push(source);
            queued[source] = true;
            while (!queue.empty()) {
                int u = queue.front();
                queue.pop();
                queued[u] = false;
                for (int i = 0; i < (int)graph[u].size(); ++i) {
                    const Edge& edge = graph[u][i];
                    if (!edge.cap) continue;
                    ll candidate = distance[u] + edge.cost;
                    if (candidate >= distance[edge.to]) continue;
                    distance[edge.to] = candidate;
                    parent_v[edge.to] = u;
                    parent_e[edge.to] = i;
                    if (!queued[edge.to]) {
                        queued[edge.to] = true;
                        queue.push(edge.to);
                    }
                }
            }
            if (distance[sink] == INF) break;
            int add = required - sent;
            for (int v = sink; v != source; v = parent_v[v])
                add = min(add,
                    graph[parent_v[v]][parent_e[v]].cap);
            for (int v = sink; v != source; v = parent_v[v]) {
                Edge& edge = graph[parent_v[v]][parent_e[v]];
                cost += ll(add) * edge.cost;
                edge.cap -= add;
                graph[v][edge.rev].cap += add;
            }
            sent += add;
        }
        return {sent, cost};
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m, required_flow;
    cin >> n >> m >> required_flow;
    int source_vertex, sink_vertex;
    cin >> source_vertex >> sink_vertex;
    --source_vertex;
    --sink_vertex;

    int super_source = n;
    int super_sink = n + 1;
    MinCostFlow network(n + 2);
    vector<int> balance(n, 0);
    ll mandatory_cost = 0;

    auto add_bounded = [&](int u, int v, int low,
                           int high, ll cost) {
        balance[u] -= low;
        balance[v] += low;
        mandatory_cost += ll(low) * cost;
        network.add_edge(u, v, high - low, cost);
    };

    for (int i = 0; i < m; ++i) {
        int u, v, low, high;
        ll cost;
        cin >> u >> v >> low >> high >> cost;
        add_bounded(u - 1, v - 1, low, high, cost);
    }
    add_bounded(sink_vertex, source_vertex,
                required_flow, required_flow, 0);

    int required = 0;
    for (int vertex = 0; vertex < n; ++vertex) {
        if (balance[vertex] > 0) {
            network.add_edge(super_source, vertex,
                             balance[vertex], 0);
            required += balance[vertex];
        } else if (balance[vertex] < 0) {
            network.add_edge(vertex, super_sink,
                             -balance[vertex], 0);
        }
    }

    auto [sent, extra_cost] =
        network.flow(super_source, super_sink, required);
    if (sent != required)
        cout << "IMPOSSIBLE\n";
    else
        cout << mandatory_cost + extra_cost << '\n';
}
